README.md
[![GitHub Release][release-img]][release]
[![GitHub Build Actions][build-action-img]][actions]
[![GitHub Release Actions][release-action-img]][actions]
[![Coverage Status][cov-img]][cov]
[![Go Report Card][report-card-img]][report-card]
[![License][license-img]][license]
![Docker Pulls / Aqua][docker-pulls-aqua]
![Docker Pulls / Harbor][docker-pulls-harbor]

# Harbor Scanner Adapter for Trivy

The Harbor [Scanner Adapter][harbor-pluggable-scanners] for [Trivy][trivy] is a service that translates
the [Harbor][harbor] scanning API into Trivy commands and allows Harbor to use Trivy for providing vulnerability
reports on images stored in Harbor registry as part of its vulnerability scan feature.

## TOC
- [Version Matrix](#version-matrix)
- [Deployment](#deployment)
  - [Harbor >= 2.0 on Kubernetes](#harbor--20-on-kubernetes)
  - [Harbor 1.10 on Kubernetes](#harbor-110-on-kubernetes)
- [Configuration](#configuration)
- [Documentation](#documentation)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)

## Version Matrix

The following matrix indicates the version of Trivy and Trivy adapter installed in each Harbor
[release](https://github.com/goharbor/harbor/releases).

| Harbor                  | Trivy Adapter | Trivy                    |
|-------------------------|---------------|--------------------------|
| -                       | v0.19.0       | [v0.17.2][trivy v0.17.2] |
| [v2.2.2][harbor v2.2.2] | v0.18.0       | [v0.16.0][trivy v0.16.0] |
| [v2.2.1][harbor v2.2.1] | v0.18.0       | [v0.16.0][trivy v0.16.0] |
| [v2.2.0][harbor v2.2.0] | v0.18.0       | [v0.16.0][trivy v0.16.0] |
| [v2.1.5][harbor v2.1.5] | v0.14.1       | [v0.9.2][trivy v0.9.2]   |
| [v2.1.0][harbor v2.1.0] | v0.14.1       | [v0.9.2][trivy v0.9.2]   |

[harbor v2.2.2]: https://github.com/goharbor/harbor/releases/tag/v2.2.2
[harbor v2.2.1]: https://github.com/goharbor/harbor/releases/tag/v2.2.1
[harbor v2.2.0]: https://github.com/goharbor/harbor/releases/tag/v2.2.0
[harbor v2.1.5]: https://github.com/goharbor/harbor/releases/tag/v2.1.5
[harbor v2.1.0]: https://github.com/goharbor/harbor/releases/tag/v2.1.0

[trivy v0.17.2]: https://github.com/aquasecurity/trivy/releases/tag/v0.17.2
[trivy v0.16.0]: https://github.com/aquasecurity/trivy/releases/tag/v0.16.0
[trivy v0.9.2]: https://github.com/aquasecurity/trivy/releases/tag/v0.9.2

## Deployment

### Harbor >= 2.0 on Kubernetes

In Harbor >= 2.0 Trivy can be configured as the default vulnerability scanner, therefore you can install it with the
official Harbor [Helm chart][harbor-helm-chart], where `HARBOR_CHART_VERSION` >= 1.4:

```
$ helm repo add harbor https://helm.goharbor.io
```

```
$ helm install harbor harbor/harbor \
    --create-namespace \
    --namespace harbor \
    --set clair.enabled=false \
    --set trivy.enabled=true
```

The adapter service is automatically registered under the **Interrogation Service** in the Harbor interface and
designated as the default scanner.

### Harbor 1.10 on Kubernetes

1. Generate certificate and private key files:
   ```
   $ openssl genrsa -out tls.key 2048
   $ openssl req -new -x509 \
       -key tls.key \
       -out tls.crt \
       -days 365 \
       -subj /CN=harbor-scanner-trivy.harbor
   ```
   > **NOTE:** The Common Name (CN) is the fully qualified domain name of the
   > adapter service. In this example we assumed that it is exposed as the
   > `harbor-scanner-trivy` service in the `harbor` namespace.
2. Install the `harbor-scanner-trivy` chart:
   ```
   $ helm repo add aqua https://helm.aquasec.com
   ```
   ```
   $ helm install harbor-scanner-trivy aqua/harbor-scanner-trivy \
       --namespace harbor \
       --set service.port=8443 \
       --set scanner.api.tlsEnabled=true \
       --set scanner.api.tlsCertificate="$(cat tls.crt)" \
       --set scanner.api.tlsKey="$(cat tls.key)"
   ```
3. Configure the scanner adapter in the Harbor interface.
   1. Navigate to **Interrogation Services** and click **+ NEW SCANNER**.
      ![Scanners config](docs/images/harbor_ui_scanners_config.png)
   2. Enter https://harbor-scanner-trivy.harbor:8443 as the **Endpoint** URL and click **TEST CONNECTION**.
      ![Add scanner](docs/images/harbor_ui_add_scanner.png)
   3. If everything is fine click **ADD** to save the configuration.
4. Select the **Trivy** scanner and set it as default by clicking **SET AS DEFAULT**.
   ![Set Trivy as default scanner](docs/images/harbor_ui_set_trivy_as_default_scanner.png)
   Make sure the **Default** label is displayed next to the **Trivy** scanner's name.

## Configuration

Configuration of the adapter is done via environment variables at startup.

|                  Name                     |                  Default           | Description |
|-------------------------------------------|------------------------------------|-------------|
| `SCANNER_LOG_LEVEL`                       | `info`                             | The log level of `trace`, `debug`, `info`, `warn`, `warning`, `error`, `fatal` or `panic`. The standard logger logs entries with that level or anything above it. |
| `SCANNER_API_SERVER_ADDR`                 | `:8080`                            | Binding address for the API server                                                   |
| `SCANNER_API_SERVER_TLS_CERTIFICATE`      | N/A                                | The absolute path to the x509 certificate file                                       |
| `SCANNER_API_SERVER_TLS_KEY`              | N/A                                | The absolute path to the x509 private key file                                       |
| `SCANNER_API_SERVER_CLIENT_CAS`           | N/A                                | A list of absolute paths to x509 root certificate authorities that the api use if required to verify a client certificate |
| `SCANNER_API_SERVER_READ_TIMEOUT`         | `15s`                              | The maximum duration for reading the entire request, including the body              |
| `SCANNER_API_SERVER_WRITE_TIMEOUT`        | `15s`                              | The maximum duration before timing out writes of the response                        |
| `SCANNER_API_SERVER_IDLE_TIMEOUT`         | `60s`                              | The maximum amount of time to wait for the next request when keep-alives are enabled |
| `SCANNER_TRIVY_CACHE_DIR`                 | `/home/scanner/.cache/trivy`       | Trivy cache directory                                                                |
| `SCANNER_TRIVY_REPORTS_DIR`               | `/home/scanner/.cache/reports`     | Trivy reports directory                                                              |
| `SCANNER_TRIVY_DEBUG_MODE`                | `false`                            | The flag to enable or disable Trivy debug mode                                       |
| `SCANNER_TRIVY_VULN_TYPE`                 | `os,library`                       | Comma-separated list of vulnerability types. Possible values are `os` and `library`. |
| `SCANNER_TRIVY_SEVERITY`                  | `UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL` | Comma-separated list of vulnerabilities severities to be displayed                   |
| `SCANNER_TRIVY_IGNORE_UNFIXED`            | `false`                            | The flag to display only fixed vulnerabilities                                       |
| `SCANNER_TRIVY_IGNORE_POLICY`             | ``                                 | The path for the Trivy ignore policy OPA Rego file                                        |
| `SCANNER_TRIVY_SKIP_UPDATE`               | `false`                            | The flag to enable or disable [Trivy DB][trivy-db] downloads from GitHub             |
| `SCANNER_TRIVY_GITHUB_TOKEN`              | N/A                                | The GitHub access token to download [Trivy DB][trivy-db] (see [GitHub rate limiting][gh-rate-limit]) |
| `SCANNER_TRIVY_INSECURE`                  | `false`                            | The flag to skip verifying registry certificate                                      |
| `SCANNER_STORE_REDIS_NAMESPACE`           | `harbor.scanner.trivy:store`       | The namespace for keys in the Redis store                                            |
| `SCANNER_STORE_REDIS_SCAN_JOB_TTL`        | `1h`                               | The time to live for persisting scan jobs and associated scan reports                |
| `SCANNER_JOB_QUEUE_REDIS_NAMESPACE`       | `harbor.scanner.trivy:job-queue`   | The namespace for keys in the scan jobs queue backed by Redis                        |
| `SCANNER_JOB_QUEUE_WORKER_CONCURRENCY`    | `1`                                | The number of workers to spin-up for the scan jobs queue                             |
| `SCANNER_REDIS_URL`                       | `redis://harbor-harbor-redis:6379` | The Redis server URI. The URI supports schemas to connect to a standalone Redis server, i.e. `redis://:password@standalone_host:port/db-number` and Redis Sentinel deployment, i.e. `redis+sentinel://:password@sentinel_host1:port1,sentinel_host2:port2/monitor-name/db-number`. |
| `SCANNER_REDIS_POOL_MAX_ACTIVE`           | `5`                                | The max number of connections allocated by the Redis connection pool                 |
| `SCANNER_REDIS_POOL_MAX_IDLE`             | `5`                                | The max number of idle connections in the Redis connection pool                      |
| `SCANNER_REDIS_POOL_IDLE_TIMEOUT`         | `5m`                               | The duration after which idle connections to the Redis server are closed. If the value is zero, then idle connections are not closed. |
| `SCANNER_REDIS_POOL_CONNECTION_TIMEOUT`   | `1s`                               | The timeout for connecting to the Redis server                                       |
| `SCANNER_REDIS_POOL_READ_TIMEOUT`         | `1s`                               | The timeout for reading a single Redis command reply                                 |
| `SCANNER_REDIS_POOL_WRITE_TIMEOUT`        | `1s`                               | The timeout for writing a single Redis command.                                      |
| `HTTP_PROXY`                              | N/A                                | The URL of the HTTP proxy server                                                     |
| `HTTPS_PROXY`                             | N/A                                | The URL of the HTTPS proxy server                                                    |
| `NO_PROXY`                                | N/A                                | The URLs that the proxy settings do not apply to                                     |

## Documentation

- [Architecture](./docs/ARCHITECTURE.md) - architectural decisions behind designing harbor-scanner-trivy.
- [Releases](./docs/RELEASES.md) - how to release a new version of harbor-scanner-trivy.

## Troubleshooting

### Error: database error: --skip-update cannot be specified on the first run

If you set the value of the `SCANNER_TRIVY_SKIP_UPDATE` to `true`, make sure that you download the Trivy DB
from [GitHub][trivy-db] and mount it in the `/home/scanner/.cache/trivy/db/trivy.db` path.

### Error: failed to list releases: GET https://api.github.com/repos/aquasecurity/trivy-db/releases: 403 API rate limit exceeded

Trivy DB downloads from GitHub are subject to [rate limiting][gh-rate-limit]. Make sure that the Trivy DB is mounted
and cached in the `/home/scanner/.cache/trivy/db/trivy.db` path. If, for any reason, it's not enough you can set the
value of the `SCANNER_TRIVY_GITHUB_TOKEN` environment variable (authenticated requests get a higher rate limit).

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull
requests.

[release-img]: https://img.shields.io/github/release/aquasecurity/harbor-scanner-trivy.svg?logo=github
[release]: https://github.com/aquasecurity/harbor-scanner-trivy/releases
[build-action-img]: https://github.com/aquasecurity/harbor-scanner-trivy/workflows/build/badge.svg
[release-action-img]: https://github.com/aquasecurity/harbor-scanner-trivy/workflows/release/badge.svg
[actions]: https://github.com/aquasecurity/harbor-scanner-trivy/actions
[cov-img]: https://codecov.io/github/aquasecurity/harbor-scanner-trivy/branch/main/graph/badge.svg
[cov]: https://codecov.io/github/aquasecurity/harbor-scanner-trivy
[report-card-img]: https://goreportcard.com/badge/github.com/aquasecurity/harbor-scanner-trivy
[report-card]: https://goreportcard.com/report/github.com/aquasecurity/harbor-scanner-trivy
[docker-pulls-aqua]: https://img.shields.io/docker/pulls/aquasec/harbor-scanner-trivy?logo=docker&label=docker%20pulls%20%2F%20aquasec
[docker-pulls-harbor]: https://img.shields.io/docker/pulls/goharbor/trivy-adapter-photon?logo=docker&label=docker%20pulls%20%2F%20goharbor
[license-img]: https://img.shields.io/github/license/aquasecurity/harbor-scanner-trivy.svg
[license]: https://github.com/aquasecurity/harbor-scanner-trivy/blob/main/LICENSE

[harbor]: https://github.com/goharbor/harbor
[harbor-helm-chart]: https://github.com/goharbor/harbor-helm
[trivy]: https://github.com/aquasecurity/trivy
[trivy-db]: https://github.com/aquasecurity/trivy-db
[latest-release-url]: https://hub.docker.com/r/aquasec/harbor-scanner-trivy/tags
[harbor-pluggable-scanners]: https://github.com/goharbor/community/blob/master/proposals/pluggable-image-vulnerability-scanning_proposal.md
[gh-rate-limit]: https://github.com/aquasecurity/trivy#github-rate-limiting


docs/RELEASES.md
# Releases

This document is intended for maintainers only.

## TOC

- [Prerequisites](#prerequisites)
- [Build and test the next release locally](#build-and-test-the-next-release-locally)
- [Release a new version](#release-a-new-version)
- [Release artifacts](#release-artifacts)

## Prerequisites

1. Install [GoReleaser](https://goreleaser.com/) or use it as curl bash piping:
   ```
   $ brew install goreleaser/tap/goreleaser
   $ goreleaser -v
   ```
   ```
   $ curl -sL https://git.io/goreleaser | bash -s -- -v
   ```
2. Fork and clone this repository and then add the `upstream` remote repository:
   ```
   $ git remote -v
   origin    git@github.com:<YOUR_GITHUB_USERNAME>/harbor-scanner-trivy.git (fetch)
   origin    git@github.com:<YOUR_GITHUB_USERNAME>/harbor-scanner-trivy.git (push)
   upstream  git@github.com:aquasecurity/harbor-scanner-trivy.git (fetch)
   upstream  git@github.com:aquasecurity/harbor-scanner-trivy.git (push)
   ```
3. Docker client connected to a Docker host:
   ```
   $ docker info
   ```

### Environment

GoReleaser requires the following environment variables to be set.

| Environment Variable | Description |
|----------------------|-------------|
| `GITHUB_TOKEN`       | GitHub API token with the `repo` scope to deploy the artifacts to GitHub |
| `DOCKERHUB_USER`     | DockerHub username |
| `DOCKERHUB_TOKEN`    | DockerHub access token to push images |

These can be stored as secrets in GitHub repository settings.

## Build and test the next release locally

1. Make sure that your fork's `main` branch is up to date with `upstream/main` and your working tree is clean.
2. Run unit tests and make sure that they're passing:
   ```
   $ make test
   ```
3. Perform a dry run to test everything before doing a release for real. Notice the `--skip-publish` flag, which
   instructs GoReleaser to only build and package things:
   ```
   $ goreleaser --snapshot --skip-publish --rm-dist
   ```
4. Make sure that the Docker image was built successfully:
   ```
   $ docker image inspect "docker.io/aquasec/harbor-scanner-trivy:$CURRENT_VERSION-next"
   ```
   where `CURRENT_VERSION` corresponds to the latest release tag, e.g. `v0.1.0` or equals `v0.0.0` if you're releasing
   for the first time.
5. You can even try running the container to be more confident with new release:
   ```
   $ docker container run --rm -p 8080:8080 "docker.io/aquasec/harbor-scanner-trivy:$CURRENT_VERSION-next"
   ```

## Release a new version

1. If everything is fine so far create an annotated git tag and push it to the `upstream` repository to actually
   trigger the release build:
   ```
   $ git tag -a $NEW_VERSION -m "Release $NEW_VERSION"
   $ git push upstream $NEW_VERSION
   ```
   where `NEW_VERSION` adheres to semantic versioning, e.g. `v0.2.0`.
2. Check that Travis CI scheduled a build job that corresponds to `NEW_VERSION`. Make sure that the job exited with 0 status code.

## Release artifacts

1. Make sure that GoReleaser uploaded artifacts to GitHub [releases](https://github.com/aquasecurity/harbor-scanner-trivy/releases) page.
2. Make sure that GoReleaser pushed new tag `NEW_VERSION` to Docker Hub [repository](https://hub.docker.com/r/aquasec/harbor-scanner-trivy/tags).


pkg/etc/config.go
package etc

import (
	"os"
	"time"

	"github.com/aquasecurity/harbor-scanner-trivy/pkg/harbor"
	"github.com/caarlos0/env/v6"
	"github.com/sirupsen/logrus"
)

type BuildInfo struct {
	Version string
	Commit  string
	Date    string
}

type Config struct {
	API        API
	Trivy      Trivy
	RedisStore RedisStore
	JobQueue   JobQueue
	RedisPool  RedisPool
}

type Trivy struct {
	CacheDir      string `env:"SCANNER_TRIVY_CACHE_DIR" envDefault:"/home/scanner/.cache/trivy"`
	ReportsDir    string `env:"SCANNER_TRIVY_REPORTS_DIR" envDefault:"/home/scanner/.cache/reports"`
	DebugMode     bool   `env:"SCANNER_TRIVY_DEBUG_MODE" envDefault:"false"`
	VulnType      string `env:"SCANNER_TRIVY_VULN_TYPE" envDefault:"os,library"`
	Severity      string `env:"SCANNER_TRIVY_SEVERITY" envDefault:"UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL"`
	IgnoreUnfixed bool   `env:"SCANNER_TRIVY_IGNORE_UNFIXED" envDefault:"false"`
	IgnorePolicy  string `env:"SCANNER_TRIVY_IGNORE_POLICY"`
	SkipUpdate    bool   `env:"SCANNER_TRIVY_SKIP_UPDATE" envDefault:"false"`
	GitHubToken   string `env:"SCANNER_TRIVY_GITHUB_TOKEN"`
	Insecure      bool   `env:"SCANNER_TRIVY_INSECURE" envDefault:"false"`
}

type API struct {
	Addr           string        `env:"SCANNER_API_SERVER_ADDR" envDefault:":8080"`
	TLSCertificate string        `env:"SCANNER_API_SERVER_TLS_CERTIFICATE"`
	TLSKey         string        `env:"SCANNER_API_SERVER_TLS_KEY"`
	ClientCAs      []string      `env:"SCANNER_API_SERVER_CLIENT_CAS"`
	ReadTimeout    time.Duration `env:"SCANNER_API_SERVER_READ_TIMEOUT" envDefault:"15s"`
	WriteTimeout   time.Duration `env:"SCANNER_API_SERVER_WRITE_TIMEOUT" envDefault:"15s"`
	IdleTimeout    time.Duration `env:"SCANNER_API_SERVER_IDLE_TIMEOUT" envDefault:"60s"`
}

func (c *API) IsTLSEnabled() bool {
	return c.TLSCertificate != "" && c.TLSKey != ""
}

type RedisStore struct {
	Namespace  string        `env:"SCANNER_STORE_REDIS_NAMESPACE" envDefault:"harbor.scanner.trivy:data-store"`
	ScanJobTTL time.Duration `env:"SCANNER_STORE_REDIS_SCAN_JOB_TTL" envDefault:"1h"`
}

type JobQueue struct {
	Namespace         string `env:"SCANNER_JOB_QUEUE_REDIS_NAMESPACE" envDefault:"harbor.scanner.trivy:job-queue"`
	WorkerConcurrency int    `env:"SCANNER_JOB_QUEUE_WORKER_CONCURRENCY" envDefault:"1"`
}

type RedisPool struct {
	URL               string        `env:"SCANNER_REDIS_URL" envDefault:"redis://localhost:6379"`
	MaxActive         int           `env:"SCANNER_REDIS_POOL_MAX_ACTIVE" envDefault:"5"`
	MaxIdle           int           `env:"SCANNER_REDIS_POOL_MAX_IDLE" envDefault:"5"`
	IdleTimeout       time.Duration `env:"SCANNER_REDIS_POOL_IDLE_TIMEOUT" envDefault:"5m"`
	ConnectionTimeout time.Duration `env:"SCANNER_REDIS_POOL_CONNECTION_TIMEOUT" envDefault:"1s"`
	ReadTimeout       time.Duration `env:"SCANNER_REDIS_POOL_READ_TIMEOUT" envDefault:"1s"`
	WriteTimeout      time.Duration `env:"SCANNER_REDIS_POOL_WRITE_TIMEOUT" envDefault:"1s"`
}

func GetLogLevel() logrus.Level {
	if value, ok := os.LookupEnv("SCANNER_LOG_LEVEL"); ok {
		level, err := logrus.ParseLevel(value)
		if err != nil {
			return logrus.InfoLevel
		}
		return level
	}
	return logrus.InfoLevel
}

func GetConfig() (Config, error) {
	var cfg Config
	err := env.Parse(&cfg)
	if err != nil {
		return cfg, err
	}

	if _, ok := os.LookupEnv("SCANNER_TRIVY_DEBUG_MODE"); !ok {
		if GetLogLevel() == logrus.DebugLevel {
			cfg.Trivy.DebugMode = true
		}
	}

	return cfg, nil
}

func GetScannerMetadata() harbor.Scanner {
	version, ok := os.LookupEnv("TRIVY_VERSION")
	if !ok {
		version = "Unknown"
	}
	return harbor.Scanner{
		Name:    "Trivy",
		Vendor:  "Aqua Security",
		Version: version,
	}
}
